test/support/factories.ex

```diff
@@ -77,23 +77,45 @@ defmodule Lightning.Factories do
 
   def log_line_factory do
     %Lightning.Invocation.LogLine{
+      id: Ecto.UUID.generate(),
       message: sequence(:log_line, &"somelog#{&1}"),
       timestamp: build(:timestamp)
     }
   end
 
-  def attempt_factory() do
+  def attempt_factory do
     %Lightning.Attempt{
       id: fn -> Ecto.UUID.generate() end
     }
   end
 
+  def attempt_with_dependencies_factory do
+    struct!(
+      attempt_factory(),
+      %{
+        created_by: build(:user),
+        work_order: build(:workorder),
+        dataclip: build(:dataclip),
+        starting_job: build(:job)
+      }
+    )
```

Comment:
I think it's preferable to use the `build` function to a create a new factory (from another one). While in this case it probably has exactly the same result but there may be other behavior we may be missing out on.

Consider writing this like:

```elixir
  def attempt_with_dependencies_factory do
    build(:attempt, 
        created_by: build(:user),
        work_order: build(:workorder),
        dataclip: build(:dataclip),
        starting_job: build(:job)
    )
  end
```

But tbh, I generally err on the side of avoiding adding variations that don't describe it's purpose - like could these fields (maybe no `created_by`) be put into the attempt factory? Or perhaps the repetition you're trying to avoid can be moved inside the test?

Would appreciate your thoughts, I'm just weary of new fixtures/factories that end up possibly living forever..

Reply:
@stuartc Thanks for the feedback! I originally had them in the individual tests, but then I ended up repeating the same helper function across 3 tests (`accounts`, `jobs` and `attempts/queue`) which is why I thought to it made sense to have them in a single place and I thought the factories may be the place to go. Happy to reverse tack, especially seeing as there is one less implementation now that I have removed the changes to `jobs_test.ex`.

My preference would be to have a factory that is 'batteries included`  along the lines of 'if you call this it will always produce something even if you have to modify that something to get what you want' but that is religion not science and I hear you re: factory sprawl - will ponder how strong my religion is :).

As for the use of `struct!` - I got that from the docs: https://hexdocs.pm/ex_machina/readme.html#overview - so I am happy to switch to exemplar you provided (I don't have a horse in this race)...although the example from the doc does seem DRYer?

Reply:
This is a tough question, you're right - we don't want factory sprawl; but we also don't want factories that are added by default to the main factories for one for a handful of cases that end up (not in this case) getting so dense people just make new ones in order to avoid breaking/understanding whats already there.

I'm happy to keep working on this, like we try and add a few more 'scenario' based factories - you can see I've played with the idea of extending factories (`with_[job|trigger|edge]`) forcing the scenarios to live in the tests but the factory implementation to live in the factories module.

Hmm interesting! Yeah I see the ex_machina library, yeah then whatever! :man_shrugging: hehe.